### PR TITLE
4 application server

### PR DESCRIPTION
**Pull request by MartianDill**
pull request resolves issue #4

### CODE REVIEW

**Comment by Misha999777** [comment]
@MartianDill are you still working on this PR?
I'm curious because it was opened and then marked as a draft.

**Comment by MartianDill** [comment]
> @MartianDill are you still working on this PR? I'm curious because it was opened and then marked as a draft.

Yes, I am. Just firstly I created Enum and then couldn't figure out how to create a working CategoryController. I will add a commit with what I have

**Comment by Misha999777** [comment]
> > @MartianDill are you still working on this PR? I'm curious because it was opened and then marked as a draft.
> 
> Yes, I am. Just firstly I created Enum and then couldn't figure out how to create a working CategoryController. I will add a commit with what I have

As we decided to make categories static we don't need a repository and a controller for them.
A repository and a controller are only needed to manipulate some dynamic data (like we use them for manipulating Application entities).
Just add a field reflecting the category to the Application entity and it would be it.

**Comment by Misha999777** [comment]
> > @MartianDill are you still working on this PR? I'm curious because it was opened and then marked as a draft.
> 
> Yes, I am. Just firstly I created Enum and then couldn't figure out how to create a working CategoryController. I will add a commit with what I have

As we decided to make categories static we don't need a repository and a controller for them.
A repository and a controller are only needed to manipulate some dynamic data (like we use them for manipulating Application entities).
Just add a field reflecting the category to the Application entity and it would be it.

File: src/main/java/cmps/charityservice/model/Category.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package cmps.charityservice.model;
+
+public enum Category {
+    ARMY(1),
+    REFUGEES(2),
+    ANIMALS(3),
+    ELDERY(4),
```
Comment (Misha999777) [inline src/main/java/cmps/charityservice/model/Category.java:7]: ELDERLY

File: src/main/java/cmps/charityservice/model/Category.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    DISABLITY(5),
```
Comment (Misha999777) [inline src/main/java/cmps/charityservice/model/Category.java:8]: DISABILITY

File: src/main/java/cmps/charityservice/model/Category.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    ONKOLOGY(6),
```
Comment (Misha999777) [inline src/main/java/cmps/charityservice/model/Category.java:9]: ONCOLOGY

File: src/main/java/cmps/charityservice/model/Category.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    ORPHANS(7),
+    HOMELESS(8),
+    EKOORG(9),     //EKO ORGANIZATIONS
+    COMMORG(10),    //COMMUNITY ORGANIZATIONS
+    INTORG(11);     //INTERNATIONAL ORGANIZATIONS
+
+    private final int id;
+
+    Category(int id) {
```
Comment (Misha999777) [inline src/main/java/cmps/charityservice/model/Category.java:18]: Replace with `@RequiredArgsConstructor` as we use Lombok

File: src/main/java/cmps/charityservice/model/Category.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        this.id = id;
+    }
+
+    public int getId() {
```
Comment (Misha999777) [inline src/main/java/cmps/charityservice/model/Category.java:22]: Replace with `@Getter` as we use Lombok

File: src/main/java/cmps/charityservice/controller/ApplicationController.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+    @DeleteMapping("/{id}")
+    @Secured("ROLE_ADMIN")
+    @Transactional
+    public void createApplication(@PathVariable String id) {
+        applicationRepository.deleteById(id);
+    }
+
+    @GetMapping("/{id}/category")
+    public Category getCategory(@PathVariable String id) {
```
Comment (Misha999777) [inline src/main/java/cmps/charityservice/controller/ApplicationController.java:70]: I don't think we need this method as Category is already available in the Application itself

File: src/main/java/cmps/charityservice/controller/ApplicationController.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                                         @RequestBody Application application) {
+        application.setId(id);
+
+        return applicationRepository.save(application);
+    }
+
+    @DeleteMapping("/{id}")
+    @Secured("ROLE_ADMIN")
+    @Transactional
+    public void createApplication(@PathVariable String id) {
```
Comment (Misha999777) [inline src/main/java/cmps/charityservice/controller/ApplicationController.java:65]: My fault, please rename to `deleteApplication`

File: src/main/java/cmps/charityservice/controller/ApplicationController.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+                                         @RequestBody Application application) {
+
+        application.setAuthorId(principal.getName());
+        application.setStatus(Status.PENDING);
+        application.setCategory(application.getCategory());
+
+        return applicationRepository.save(application);
+    }
+
+    @PostMapping("/{id}")
```
Comment (Misha999777) [inline src/main/java/cmps/charityservice/controller/ApplicationController.java:53]: My fault, should be `@PutMapping("/{id}")` instead

File: src/main/java/cmps/charityservice/controller/ApplicationController.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+@RestController
+@RequestMapping("/applications")
+@CrossOrigin
+@RequiredArgsConstructor
+public class ApplicationController {
+
+    private final ApplicationRepository applicationRepository;
+
+    @GetMapping
+    public List<Application> getAll(JwtAuthenticationToken principal,
```
Comment (Misha999777) [inline src/main/java/cmps/charityservice/controller/ApplicationController.java:27]: Let's add the ability to query Applications by a specific Category

**Review by Misha999777 — COMMENTED** [review]
LGTM, except for comments

**Comment by MartianDill** [comment]
@Misha999777 I've added all requested fixes, but not sure if the if-else in ApplicationController should be used
```java
public List<Application> getAll(JwtAuthenticationToken principal,
...
...
if (category != null) {
       return applicationRepository.findAllByStatusAndCategory(search, category, PageRequest.of(page, size));
} else {
       return applicationRepository.findAllByStatus(search, PageRequest.of(page, size));
}
```

**Comment by Misha999777** [comment]
> @Misha999777 I've added all requested fixes, but not sure if the if-else in ApplicationController is the right thing to do
> 
> ```java
> public List<Application> getAll(JwtAuthenticationToken principal,
> ...
> ...
> if (category != null) {
>        return applicationRepository.findAllByStatusAndCategory(search, category, PageRequest.of(page, size));
> } else {
>        return applicationRepository.findAllByStatus(search, PageRequest.of(page, size));
> }
> ```

It’s a good way to do this

**Review by Misha999777 — APPROVED** [review]
LGTM

**Pull request merged**